File: 2022-hs/src/Day1.hs
module Day1
    ( part1
    , part2
    , example
    ) where

import Data.List
import Data.List.Split (splitOn)

example = "1000\n\
          \2000\n\
          \3000\n\
          \\n\
          \4000\n\
          \\n\
          \5000\n\
          \6000\n\
          \\n\
          \7000\n\
          \8000\n\
          \9000\n\
          \\n\
          \10000"

elfSums = map (sum . map read . lines) . splitOn "\n\n"
part1 = maximum . elfSums
part2 = sum . take 3 . reverse . sort . elfSums

File: 2022/src/day1.rs
use itertools::Itertools;

fn parse(input: &str) -> impl Iterator<Item = u64> + '_ {
    input
        .split("\n\n")
        .map(|elf| elf.lines().map(|l| l.parse::<u64>().unwrap()).sum())
}

pub fn part1(input: &str) -> u64 {
    parse(input).max().unwrap()
}

pub fn part2(input: &str) -> u64 {
    parse(input).sorted().rev().take(3).sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    const EXAMPLE_INPUT: &'static str = "1000
2000
3000

4000

5000
6000

7000
8000
9000

10000";

    #[test]
    fn p1_example() {
        assert_eq!(part1(EXAMPLE_INPUT), 24000);
    }

    #[test]
    fn p1_input() {
        let input = std::fs::read_to_string("input/day1").expect("reading input file");
        assert_eq!(part1(&input), 69883);
    }

    #[test]
    fn p2_example() {
        assert_eq!(part2(EXAMPLE_INPUT), 45000);
    }

    #[test]
    fn p2_input() {
        let input = std::fs::read_to_string("input/day1").expect("reading input file");
        assert_eq!(part2(&input), todo!());
    }
}
